Move some commas around (#71)
soupi/learn-haskell-blog-generator: src/05-glue/02-io.md

or so we would expect if this was a strict language. In a strict language,
we expect the order of operations to happen from the top down.

But in a lazy language, we don't evaluate an expression until
it is needed, and so neither `result1` nor `result2` are evaluated
until we wish to print the result of subtracting one from the other.
And then, when we try to evaluate `-`, it evaluates the two arguments
from left to right, so we first evaluate `result2`.

Evaluating `result2` with substitution means replacing occurrences of `n`
with the input `2`, and then evaluate the top-level function (`+`), which is a
primitive function. We then evaluate its arguments, `readIntFromStdin`
and then `n`; at this point *we are reading the first integer from the stdin*.

After calculating the result, we can move to evaluate `result1`, which
*will read the second integer from stdin*. This is the
complete opposite of what we wanted!

Issues like these make lazy evaluation hard to work within the presence of
**side effects** - when the evaluation of an expression *can affect or be affected
by the outside world*, this includes reading/writing from mutable memory
or performing I/O operations.

> We call functions with side-effects, such as `addWithInput`, **impure functions**.

Reviewer: ```suggestion
We call functions that have side-effects, such as `addWithInput`, **impure functions**.
```

have **no side effects**. Pure functions will *always* return the same output (given the same input)
and **evaluating pure expressions is deterministic**.

But now, how can we do I/O operations? There are many possible solutions.

For Haskell, it was decided to design an interface
with an accompanied type called `IO`. `IO`'s interface will force a distinction

> from non-I/O expressions and will also require that to *combine*

Reviewer: ```suggestion
from non-I/O expressions and will also require that in order to *combine*
```

multiple `IO` operations, we will have to **specify the order of the operations**.

## IO

`IO` is an opaque type, like our `Html` type, in which we hide its internal
representation from the user behind an interface. But in this case, `IO` is a

> built-in type hidden by the Haskell language rather than a module.

Reviewer: ```suggestion
built-in type and its internals are hidden by the Haskell language rather than a module.
```

```

<!-- You can find more useful IO functions and operations in the
[System.IO](https://hackage.haskell.org/package/base-4.16.4.0/docs/System-IO.html) module,
which is shipped with `base`. -->

Notice that each function returns an `IO <something>`, but what does that mean?

The meaning behind `IO a` is that it is a *description of a program (or subroutine)
that, when executed, will produce some value of type `a`, and may do some I/O effects
during execution*.

Executing an `IO a` is different from evaluating it.
Evaluating an `IO a` expression is pure - the **evaluation** will always reduce to

> the same **description** of a program. This helps us keep the purity and equational reasoning!

Reviewer: The purity!

```suggestion
the same **description** of a program. This helps us keep purity and equational reasoning!
```

The Haskell runtime will *execute* the entry point to the program

> (the expression `main` that must have the type `IO ()`) for our IO operation

Reviewer: ```suggestion
(the expression `main` that must have the type `IO ()`). For our IO operation
```
